plugins/launcher_service_plugin/include/eosio/launcher_service_plugin/launcher_service_plugin.hpp

```diff
@@ -0,0 +1,357 @@
+      uint16_t p2p_port(const launcher_config &config, int cluster_id) const {
+         if (assigned_p2p_port) {
+            return assigned_p2p_port;
+         } else {
+            return http_port(config, cluster_id) + 1;
+         }
+      }
+      bool is_bios() const {
+         for (auto &p: producers) {
+            if (p == "eosio") return true;
+         }
+         return false;
+      }
+   };
+
+   struct cluster_def {
+      std::string                    shape = "mesh"; // "mesh" / "star" / "bridge" / "line" / "ring" / "tree"
+      int                            center_node_id = -1; // required if shape is "start" or "bridge"
+      int                            cluster_id = 0;
+      int                            node_count = 0;
+      bool                           auto_port = true; // auto port assignment
+      std::vector<node_def>          nodes;
+      std::vector<string>            extra_configs;
+      std::string                    extra_args;
+      fc::log_level                  log_level = fc::log_level::info;
+      std::map<std::string, fc::log_level> special_log_levels;
+
+      bool                           _normalized = false;
+
+      node_def &get_node_def(int id) {
+         if (!_normalized) {
+            for (int i = 0; i < node_count; ++i) { // normalize
+               bool found = false;
+               for (auto &n: nodes) {
+                  if (n.node_id == i) {
+                     found = true;
+                     break;
+                  }
+               }
+               if (found) continue;
+               node_def n;
+               n.node_id = i;
+               nodes.push_back(n);
+            }
+            _normalized = true;
+         }
+         for (auto &n: nodes) {
+            if (n.node_id == id) return n;
+         }
+         throw std::runtime_error("node config not found");
+      }
+   };
+
+   struct new_account_param {
+      chain::name                    name;
+      public_key_type                owner;
+      public_key_type                active;
+   };
+   struct create_bios_accounts_param {
+      int                            cluster_id = 0;
+      int                            node_id = 0;
+      chain::name                    creator;
+      std::vector<new_account_param> accounts;
+   };
+   struct link_auth_param {
+      int                            cluster_id = 0;
+      int                            node_id = 0;
+      chain::name                    account;
+      chain::name                    code;
+      chain::name                    type;
+      chain::name                    requirement;
+   };
+   struct unlink_auth_param {
+      int                            cluster_id = 0;
+      int                            node_id = 0;
+      chain::name                    account;
+      chain::name                    code;
+      chain::name                    type;
+   };
+   struct new_account_param_ex { // newaccount in system contract
```

Comment:
System contract specific actions should not be part of the launcher service.

Reply:
removed